File: experiments/dump_areas.pl
#!/usr/bin/perl
# Dump per-leaf areas from a .dom using Urb itself (ground truth for validating
# the Python geometry port). Run from the urb repo root with -Ilib.
# Output matches experiments/dump_areas.py: "level/idpath type area", sorted.
use strict;
use warnings;
use YAML;
use Urb::Dom;

my $dom = Urb::Dom->new;
$dom->Deserialise (YAML::LoadFile ($ARGV[0]));

my @rows;
for my $level ($dom, $dom->Levels_Above)
{
    my $lid = scalar $level->Levels_Below;
    for my $leaf ($level->Leafs)
    {
        my $type = $leaf->Type || '?';
        push @rows, sprintf ("%d/%s %s %.4f", $lid, $leaf->Id, $type, $leaf->Area);
    }
}
print join ("\n", sort @rows), "\n";

File: experiments/dump_areas.py
"""Dump per-leaf areas from a .dom using the Python geometry port.

Used to validate the port against a Perl dump from Urb (see the sibling
``dump_areas.pl``). Output: one line per leaf, ``level/idpath type area``,
sorted for a stable diff.
"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))

from homemaker import dom, geometry  # noqa: E402


def main(path: str) -> None:
    root = dom.load(path)
    rows = []
    for level_idx, lvl in enumerate(dom.levels(root)):
        for leaf in lvl.leaves():
            rows.append(f"{level_idx}/{leaf.id} {leaf.type or '?'} {geometry.area(leaf):.4f}")
    print("\n".join(sorted(rows)))


if __name__ == "__main__":
    main(sys.argv[1])
